starboard/android/shared/media_decoder.cc

```diff
@@ -463,24 +463,20 @@ bool MediaDecoder::ProcessOneInputBuffer(
   }
 
   jint status;
-  if (event.type == Event::kWriteCodecConfig) {
-    if (!drm_system_ || (drm_system_ && drm_system_->IsReady())) {
-      status = media_codec_bridge_->QueueInputBuffer(dequeue_input_result.index,
-                                                     kNoOffset, size, kNoPts,
-                                                     BUFFER_FLAG_CODEC_CONFIG);
-    } else {
-      status = MEDIA_CODEC_NO_KEY;
-    }
+  if (drm_system_ && !drm_system_->IsReady()) {
+    // Drm system initialization is asynchronous. If there's a drm system, we
+    // should wait until it's initialized to avoid errors.
+    status = MEDIA_CODEC_NO_KEY;
+  } else if (event.type == Event::kWriteCodecConfig) {
+    status = media_codec_bridge_->QueueInputBuffer(dequeue_input_result.index,
+                                                   kNoOffset, size, kNoPts,
+                                                   BUFFER_FLAG_CODEC_CONFIG);
   } else if (event.type == Event::kWriteInputBuffer) {
     jlong pts_us = input_buffer->timestamp();
     if (drm_system_ && input_buffer->drm_info()) {
```

Comment:
This code is to ensure that even when there is a drm system (i.e. the stream is encrypted), as long as the current input buffer isn't encrypted, we can still play it without waiting for the drm system.

The logic seems to be broken under the new implementation.

Reply:
https://github.com/youtube/cobalt/blob/a8bd2c377c40a01ff47c6f2c6d6cee3f067c096b/starboard/android/apk/app/src/main/java/dev/cobalt/media/MediaDrmBridge.java#L541
When instantiating a drm system, it uses an empty MediaCrypto and then opening MediaCryptoSession on a different thread. When we write any input, if we haven't receive the real MediaCrypto, MediaCodec would throw out errors even though the input is clear. So, we cannot write any input into the MediaCodec, until MediaCrypto is ready to use (which is when IsReady() returns true).

The old implementation created MediaCryptoSession synchronously in ctor, so it didn't have that problem.

Reply:
Do we know the extra delay introduced in this case?

Reply:
Compared to the implementation before pr #1337 (device provisioning was moved to a separate thread), no extra delay was added.